### ISSUE TITLE
MIROC6-tasmax.yaml fails to apply QDM biascorrection to historical with "KeyError: '2026'" 

### ISSUE THREAD

**Issue by brews** [ClimateImpactLab/downscaleCMIP6#371]
Similar to #368 but with MIROC6.

Archived workflow is here https://argo.cildc6.org/archived-workflows/default/d9d08f31-63d6-44c2-ab3a-9035b2cd39bd.

Logged error for one of the apply-qdm latitudinal slices:

```
INFO:dodola.services:Starting dodola service apply_qdm with args=(), kwargs={'simulation': 'gs://scratch-170cd6ec/d9d08f31-63d6-44c2
-ab3a-9035b2cd39bd/e2e-miroc6-tasmax-h6jwf-2520553948/preprocessed.zarr', 'qdm': 'gs://scratch-170cd6ec/d9d08f31-63d6-44c2-ab3a-9035
b2cd39bd/e2e-miroc6-tasmax-h6jwf-84168436/qdm_model.zarr', 'years': range(1950, 2100), 'variable': 'tasmax', 'out': 'gs://scratch-17
0cd6ec/d9d08f31-63d6-44c2-ab3a-9035b2cd39bd/e2e-miroc6-tasmax-h6jwf-2691629103/qdm_adjusted.zarr', 'sel_slice': None, 'isel_slice':
{'lat': slice(0, 10, None)}, 'out_zarr_region': {'lat': slice(0, 10, None)}, 'root_attrs_json_file': '/tmp/global_attrs.json', 'new_
attrs': None})
INFO:dodola.repository:Read gs://scratch-170cd6ec/d9d08f31-63d6-44c2-ab3a-9035b2cd39bd/e2e-miroc6-tasmax-h6jwf-2520553948/preprocess
ed.zarr
INFO:dodola.repository:Read gs://scratch-170cd6ec/d9d08f31-63d6-44c2-ab3a-9035b2cd39bd/e2e-miroc6-tasmax-h6jwf-84168436/qdm_model.za
rr
INFO:dodola.services:Using root attrs from /tmp/global_attrs.json
INFO:dodola.repository:Read attrs from /tmp/global_attrs.json
INFO:dodola.services:Slicing by isel_slice={'lat': slice(0, 10, None)}
Traceback (most recent call last):
  File "/opt/conda/lib/python3.9/site-packages/xarray/coding/cftimeindex.py", line 430, in _get_string_slice
    loc = self._partial_date_slice(resolution, parsed)
  File "/opt/conda/lib/python3.9/site-packages/xarray/coding/cftimeindex.py", line 415, in _partial_date_slice
    raise KeyError
KeyError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/bin/dodola", line 33, in <module>
    sys.exit(load_entry_point('dodola', 'console_scripts', 'dodola')())
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/conda/lib/python3.9/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/opt/dodola/dodola/cli.py", line 231, in apply_qdm
    services.apply_qdm(
  File "/opt/dodola/dodola/services.py", line 32, in service_logger
    func(*args, **kwargs)
  File "/opt/dodola/dodola/services.py", line 317, in apply_qdm
    adjusted_ds = adjust_quantiledeltamapping(
  File "/opt/dodola/dodola/core.py", line 101, in adjust_quantiledeltamapping
    adj = adjust_quantiledeltamapping_year(
  File "/opt/dodola/dodola/core.py", line 175, in adjust_quantiledeltamapping_year
    out = qdm.adjust(simulation, interp="nearest").sel(time=str(year))
  File "/opt/conda/lib/python3.9/site-packages/xarray/core/dataset.py", line 2474, in sel
    pos_indexers, new_indexes = remap_label_indexers(
  File "/opt/conda/lib/python3.9/site-packages/xarray/core/coordinates.py", line 421, in remap_label_indexers
    pos_indexers, new_indexes = indexing.remap_label_indexers(
  File "/opt/conda/lib/python3.9/site-packages/xarray/core/indexing.py", line 117, in remap_label_indexers
    idxr, new_idx = index.query(labels, method=method, tolerance=tolerance)
  File "/opt/conda/lib/python3.9/site-packages/xarray/core/indexes.py", line 224, in query
    indexer = index.get_loc(
  File "/opt/conda/lib/python3.9/site-packages/xarray/coding/cftimeindex.py", line 464, in get_loc
    return self._get_string_slice(key)
  File "/opt/conda/lib/python3.9/site-packages/xarray/coding/cftimeindex.py", line 432, in _get_string_slice
    raise KeyError(key)
KeyError: '2026'
```

Again, something with historical is fishy. None of the other targets/ssps appeared to have issues.

Blocking progress on run #191

**Comment by dgergel** [comment]
I'm a bit confused on this error @brews. I inspected the input zarr, which goes from 1950 - 2025. This checks out for historical, since we append 2015 - 2025 to historical for the rolling 21-yr. 

But then the `KeyError: '2026'` - why is `apply_qdm` looking for 2026 in historical to apply qdm to that year?? We should only be "applying" QDM for the historical period *through* 2014, which is the end of 2014. Maybe I'm missing something here, or you meant something slightly different in the Github issue description. 

Just want to be sure our workflow isn't actually applying QDM to 2015 - on for the historical model run.

**Comment by brews** [comment]
Yeah, I think I understand @dgergel .

Let me chase this down. I have a feeling that "historical" is running with the wrong years after we made the recent 2099 (or 2100) adjustment. 

I'll follow up.

**Comment by brews** [comment]
@dgergel Yeah, this is a bug from my PR #364. I see what's going on. So, the historical run is trying its hardest to bias-correct from 1950 - 2099 (and failing hard).

I estimate I can have this fixed by tomorrow. I'll get it fixed before redoing the runs fixed in PR #377.

I don't think we had any historical runs to completion with this PR #364 (because it bombs on trying to bias correcting non-existent data), so no worries there.

**Comment by brews** [comment]
Note to self, we're going to want to fix up `workflows/templates/qdm.yaml` to fix the changes in PR #364, but consider moving that dynamic 2099-2100 end-finding step to `workflows/templates/biascorrectdownscale.yaml`. If decide to keep it in qdm.yaml, then need to remove last-year parameter passing in biascorrectdownscale.yaml because it's misleading. We should at least try to be clear about where this first-year, last-year responsibility rests -- i.e. not smeared across multiple workflowtemplates.
